hal/src/peripherals/adc/async_api.rs

```diff
@@ -0,0 +1,81 @@
+use core::marker::PhantomData;
+
+use crate::{
+    adc::{Adc, AdcInstance, Error, Flags},
+    async_hal::interrupts::Handler,
+};
+
+use atsamd_hal_macros::hal_module;
+use embassy_sync::waitqueue::AtomicWaker;
+
+#[allow(clippy::declare_interior_mutable_const)]
+const NEW_WAKER: AtomicWaker = AtomicWaker::new();
+
+#[hal_module(any("adc-d11", "adc-d21"))]
+pub static ADC_WAKERS: [AtomicWaker; 1] = [NEW_WAKER; 1];
+#[hal_module("adc-d5x")]
+pub static ADC_WAKERS: [AtomicWaker; 2] = [NEW_WAKER; 2];
+
+/// Interrupt handler for the ADC peripheral.
+pub struct InterruptHandler<A: AdcInstance> {
+    _private: (),
+    _adc: PhantomData<A>,
+}
+
+impl<A: AdcInstance> crate::typelevel::Sealed for InterruptHandler<A> {}
+
+impl<A: AdcInstance> Handler<A::Interrupt> for InterruptHandler<A> {
+    unsafe fn on_interrupt() {
+        let mut peripherals = crate::pac::Peripherals::steal();
+        let adc = A::peripheral_reg_block(&mut peripherals);
+
+        let flags_pending = Flags::from_bits_truncate(adc.intflag().read().bits());
+        let enabled_flags = Flags::from_bits_truncate(adc.intenset().read().bits());
+
+        if enabled_flags.intersects(flags_pending) {
+            adc.intenclr().write(|w| w.bits(flags_pending.bits()));
+            // Wake up!
+            A::waker().wake();
+        }
+    }
+}
+
+impl<I: AdcInstance, F> Adc<I, F>
+where
+    F: crate::async_hal::interrupts::Binding<I::Interrupt, InterruptHandler<I>>,
+{
+    #[inline]
+    pub(super) async fn wait_flags(&mut self, flags_to_wait: Flags) -> Result<(), Error> {
+        use core::task::Poll;
+
+        // We automatically check for errors
+        let flags_to_wait = flags_to_wait | Flags::OVERRUN;
+        self.disable_interrupts(Flags::all());
```

Comment:
I'm curious why this disables all interrupts, as opposed to just `flags_to_wait` (which is what gets re-enabled)

Reply:
its to prevent a race condition as far as @jbeaurivage explained to me. Disable everything before starting the polling (Which then enables the interrupts once it is ready)

Reply:
I guess the risk is that we later want to do something with one of the other flags, `WINMON` or whatever.  In that case though, this code would probably need to be revisited either way.

I see that at least the SPI async implementation does the same thing so may as well be consistent.

Reply:
Good point @ianrrees, we could simply disable the flags we want to wait on. My thinking (and the same goes for all async implementations in the HAL), is that disabling everything can prevent previously enabled (and unused) interrupt sources from uselessly waking the task.